File: libs/server/api/src/api/model/member_model.rs
use serde::Serialize;
use std::collections::BTreeMap;

use crate::domain::Member;

use super::super::links::{
    user_href, workspace_href, workspace_member_href, workspace_members_href, Link,
};

#[derive(Debug, Clone, Serialize)]
struct RefModel {
    #[serde(rename = "_links")]
    links: BTreeMap<String, Link>,
    id: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(in crate::api) struct MemberModel {
    #[serde(rename = "_links")]
    links: BTreeMap<String, Link>,
    id: String,
    workspace: RefModel,
    user: RefModel,
    role: String,
    created_at: String,
    updated_at: String,
}

pub(in crate::api) fn member_model(user_id: &str, member: &Member) -> MemberModel {
    let workspace_id = member.workspace_id();
    let member_user_id = member.description().user.id();
    MemberModel {
        links: BTreeMap::from([
            (
                "self".to_string(),
                Link::new(workspace_member_href(
                    user_id,
                    workspace_id,
                    member.identity(),
                )),
            ),
            (
                "collection".to_string(),
                Link::new(workspace_members_href(user_id, workspace_id)),
            ),
            (
                "workspace".to_string(),
                Link::new(workspace_href(user_id, workspace_id)),
            ),
            ("user".to_string(), Link::new(user_href(member_user_id))),
        ]),
        id: member.identity().to_string(),
        workspace: RefModel {
            links: BTreeMap::from([(
                "self".to_string(),
                Link::new(workspace_href(user_id, workspace_id)),
            )]),
            id: workspace_id.to_string(),
        },
        user: RefModel {
            links: BTreeMap::from([("self".to_string(), Link::new(user_href(member_user_id)))]),
            id: member_user_id.clone(),
        },
        role: member.description().role.clone(),
        created_at: member.created_at().to_string(),
        updated_at: member.updated_at().to_string(),
    }
}

File: libs/server/api/src/api/model/member-model.ts
import { Member, Workspace } from '@evidence/server-domain';
import {
  link,
  Link,
  userHref,
  workspaceHref,
  workspaceMemberHref,
  workspaceMembersHref,
} from '../links';
import { WorkspaceModel, workspaceModel } from './workspace-model';

interface RefModel {
  _links: Record<string, Link>;
  id: string;
}

export interface MemberModel {
  _links: Record<string, Link>;
  id: string;
  workspace: RefModel;
  user: RefModel;
  role: string;
  createdAt: string;
  updatedAt: string;
}

export interface MembershipModel {
  _links: Record<string, Link>;
  id: string;
  workspace: WorkspaceModel;
  user: RefModel;
  role: string;
  createdAt: string;
  updatedAt: string;
}

export function memberModel(member: Member): MemberModel {
  const description = member.description();
  const workspaceId = description.workspace.id();
  const memberUserId = description.user.id();
  return {
    _links: memberLinks(member),
    id: member.identity(),
    workspace: {
      _links: { self: link(workspaceHref(workspaceId)) },
      id: workspaceId,
    },
    user: {
      _links: { self: link(userHref(memberUserId)) },
      id: memberUserId,
    },
    role: description.role,
    createdAt: description.createdAt,
    updatedAt: description.updatedAt,
  };
}

export function membershipModel(
  membership: Member,
  workspace: Workspace,
): MembershipModel {
  const description = membership.description();
  const memberUserId = description.user.id();
  return {
    _links: memberLinks(membership),
    id: membership.identity(),
    workspace: workspaceModel(workspace),
    user: {
      _links: { self: link(userHref(memberUserId)) },
      id: memberUserId,
    },
    role: description.role,
    createdAt: description.createdAt,
    updatedAt: description.updatedAt,
  };
}

function memberLinks(member: Member): Record<string, Link> {
  const description = member.description();
  const workspaceId = description.workspace.id();
  const memberUserId = description.user.id();
  return {
    self: link(workspaceMemberHref(workspaceId, member.identity())),
    collection: link(workspaceMembersHref(workspaceId)),
    workspace: link(workspaceHref(workspaceId)),
    user: link(userHref(memberUserId)),
  };
}
